src/uct/ib/ud/base/ud_ep.c

```diff
@@ -140,15 +141,97 @@ static ucs_status_t uct_ud_ep_free_by_timeout(uct_ud_ep_t *ep,
     return UCS_INPROGRESS;
 }
 
+static UCS_F_ALWAYS_INLINE void
+uct_ud_ep_window_release(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
+                         uct_ud_psn_t max_psn, ucs_status_t status, int is_async)
+{
+    uct_ud_send_skb_t *skb;
+
+    ucs_queue_for_each_extract(skb, &ep->tx.window, queue,
+                               UCT_UD_PSN_COMPARE(skb->neth->psn, <=, max_psn)) {
+        ucs_assert(!(skb->flags & UCT_UD_SEND_SKB_FLAG_INVALID));
+        if (ucs_likely(!(skb->flags & (UCT_UD_SEND_SKB_FLAG_RESENDING |
```

Comment:
maybe unite this and next `if` to the one single `if`:
```
if (ucs_likely(!(skb->flags & UCT_UD_SEND_SKB_FLAG_RESENDING))) {
    if (!(skb->flags & UCT_UD_SEND_SKB_FLAG_COMP)) {
        /* fast path case: completed skb without completion callback */
        uct_ud_skb_release(skb, 1);
    } else {
        ... <the body of the second if>
    }
} else {
    ... <the body of the thrid if>
}
```

Reply:
this is for optimization reason: so that fast path case could be checked by 1 branch (check 2 flags together)

Reply:
> this is for optimization reason: so that fast path case could be checked by 1 branch (check 2 flags together)

ah, I see, thank you for the explanation!